Mathlib/NumberTheory/ModularForms/JacobiTheta/Bounds.lean

```diff
@@ -0,0 +1,271 @@
+  · rw [one_pow, mul_one]
+
+lemma isBigO_atTop_F_nat_zero_sub {a : ℝ} (ha : 0 ≤ a) : ∃ p, 0 < p ∧
+    (fun t ↦ F_nat 0 a t - (if a = 0 then 1 else 0)) =O[atTop] fun t ↦ exp (-p * t) := by
+  have aux : IsBigO atTop (fun t : ℝ ↦ (1 - rexp (-π * t))⁻¹) (fun _ ↦ (1 : ℝ)) := by
+    refine ((Tendsto.const_sub _ ?_).inv₀ (by norm_num)).isBigO_one ℝ (c := ((1 - 0)⁻¹ : ℝ))
+    simpa only [neg_mul, tendsto_exp_comp_nhds_zero, tendsto_neg_atBot_iff]
+      using tendsto_id.const_mul_atTop pi_pos
+  split_ifs with h
+  · rw [h]
+    have : (fun t ↦ F_nat 0 0 t - 1) =O[atTop] fun t ↦ rexp (-π * t) / (1 - rexp (-π * t)) := by
+      apply Eventually.isBigO
+      filter_upwards [eventually_gt_atTop 0] with t ht
+      exact F_nat_zero_zero_sub_le ht
+    refine ⟨_, pi_pos, this.trans ?_⟩
+    simpa using (isBigO_refl (fun t ↦ rexp (-π * t)) _).mul aux
+  · simp_rw [sub_zero]
+    have : (fun t ↦ F_nat 0 a t) =O[atTop] fun t ↦ rexp (-π * a ^ 2 * t) / (1 - rexp (-π * t)) := by
+      apply Eventually.isBigO
+      filter_upwards [eventually_gt_atTop 0] with t ht
+      exact F_nat_zero_le ha ht
+    refine ⟨π * a ^ 2, mul_pos pi_pos (sq_pos_of_ne_zero _ h), this.trans ?_⟩
+    simp_rw [neg_mul π (a ^ 2)]
+    simpa only [mul_one] using (isBigO_refl (fun t ↦ rexp (-(π * a ^ 2) * t)) _).mul aux
+
+end k_eq_zero
+
+section k_eq_one
+/-!
+## Sum over `ℕ` with `k = 1`
+
+Here we use comparison with the series `∑ n r ^ n`, where `r = exp (-π * t)`.
+-/
+
+lemma F_nat_one_le {a : ℝ} (ha : 0 ≤ a) {t : ℝ} (ht : 0 < t) :
+    ‖F_nat 1 a t‖ ≤ rexp (-π * (a ^ 2 + 1) * t) / (1 - rexp (-π * t)) ^ 2
+      + a * rexp (-π * a ^ 2 * t) / (1 - rexp (-π * t)) := by
+  refine tsum_of_norm_bounded ?_ (f_le_g_nat 1 ha ht)
+  have h0 : rexp (-π * t) < 1 := by
+    simpa only [exp_lt_one_iff, neg_mul, neg_lt_zero] using mul_pos pi_pos ht
+  simp_rw [g_nat, pow_one, add_mul]
+  apply HasSum.add
+  · have h0' : ‖rexp (-π * t)‖ < 1 := by rwa [norm_eq_abs, abs_exp]
+    convert (hasSum_coe_mul_geometric_of_norm_lt_one h0').mul_left (exp (-π * a ^ 2 * t)) using 1
+    · ext1 n
+      rw [mul_comm (exp _), ← Real.exp_nat_mul, mul_assoc (n : ℝ), ← Real.exp_add]
+      ring_nf
+    · rw [mul_add, add_mul, mul_one, exp_add, mul_div_assoc]
+  · convert (hasSum_geometric_of_lt_one (exp_pos _).le h0).mul_left _ using 1
+    ext1 n
+    rw [← Real.exp_nat_mul, mul_assoc _ (exp _), ← Real.exp_add]
+    ring_nf
+
+lemma isBigO_atTop_F_nat_one {a : ℝ} (ha : 0 ≤ a) : ∃ p, 0 < p ∧
+    F_nat 1 a =O[atTop] fun t ↦ exp (-p * t) := by
+  suffices ∃ p, 0 < p ∧ (fun t ↦ rexp (-π * (a ^ 2 + 1) * t) / (1 - rexp (-π * t)) ^ 2
+      + a * rexp (-π * a ^ 2 * t) / (1 - rexp (-π * t))) =O[atTop] fun t ↦ exp (-p * t) by
+    let ⟨p, hp, hp'⟩ := this
+    refine ⟨p, hp, (Eventually.isBigO ?_).trans hp'⟩
+    filter_upwards [eventually_gt_atTop 0] with t ht
+    exact F_nat_one_le ha ht
+  have aux : IsBigO atTop (fun t : ℝ ↦ (1 - rexp (-π * t))⁻¹) (fun _ ↦ (1 : ℝ)) := by
```

Comment:
This is the second time this statement is needed. Extract?

Reply:
This one's maybe a bit too specific, it's very unlikely to be ever needed again after this.

Reply:
I've made it a private lemma as per your other suggestion.